File: RobotClient/controls.py
"""
The robot controlling functions
Only left(), right() and stop() will be called after a call to forward() or backward()
Only forward(), backward() and stop() will be called after a call to left() or right()
"""

# External module imports
import RPi.GPIO as GPIO
import time

################## VARIABLE DEFINITIONS ##################

# Motor Driver Pin declarations
motorA_pwm = 12 #GPIO12 (Pin 32) for Motor A PWM speed control
motorA_01 = 5 #GPIO05 (Pin 29) for Motor A 01 input
motorA_02 = 6 #GPIO06 (Pin 31) for Motor A 02 input
motorB_01 = 13 #GPIO13 (Pin 33) for Motor B 01 input
motorB_02 = 19 #GPIO19 (Pin 35) for Motor B 02 input
motorB_pwm = 26 #GPIO26 (Pin 37) for Motor B PWM speed control
stdby = 21 #GPIO21 (Pin 40) for Standby pin on Motor Driver

# PWM Constants
dc_max = 100
dc_min = 0
dc_speed_inc = 60
dc_inc = 20

# pwm frequency for setup
pwm_freq = 50 

# Default pin modes
dc_pwmA = dc_min
dc_pwmB = dc_min
last_dc_pwmA = dc_pwmA
last_dc_pwmB = dc_pwmB
motorA_01_state = GPIO.LOW
motorA_02_state = GPIO.LOW
motorB_01_state = GPIO.LOW
motorB_02_state = GPIO.LOW

# Last call states 
FORWARD = 1
BACKWARD = 2
RIGHT = 3
LEFT = 4
STOP = 0

# Default lastCall to STOP
lastCall = STOP

######################### INITIALIZER ######################

# initializes pins (be sure to call before using other methods)
def init(): 
    # Pin Setup:
    GPIO.setmode(GPIO.BCM) # Broadcom pin-numbering scheme

    GPIO.setup(motorA_pwm, GPIO.OUT) # LED pin set as output
    GPIO.setup(motorA_01, GPIO.OUT) # PWM pin set as output
    GPIO.setup(motorA_02, GPIO.OUT) # PWM pin set as output
    GPIO.setup(motorB_01, GPIO.OUT) # PWM pin set as output
    GPIO.setup(motorB_02, GPIO.OUT) # PWM pin set as output
    GPIO.setup(motorB_pwm, GPIO.OUT) # PWM pin set as output
    GPIO.setup(stdby, GPIO.OUT) # Standby pin on motor driver needs to be pulled high 

    global pwmA
    pwmA = GPIO.PWM(motorA_pwm, pwm_freq) #Initialize PWM on motorA_pwm 100Hz frequency
    global pwmB
    pwmB  = GPIO.PWM(motorB_pwm, pwm_freq) #Initialize PWM on motorA_pwm 100Hz frequency

    GPIO.output(stdby, GPIO.HIGH)

    pwmA.start(dc_pwmA)
    pwmB.start(dc_pwmB)
    stop()

################ FORWARD and BACKWARD METHODS #################

# makes the robot go forward 
def forward():
    """Make the robot keep moving forward"""
    print("forward")
    
    if lastCall == FORWARD: 
        #Speed incrementing
        if dc_pwmA + dc_speed_inc <= dc_max: 
            dc_pwmA += dc_speed_inc
        else:
            dc_pwmA = dc_max
        
        if dc_pwmB + dc_speed_inc <= dc_max: 
            dc_pwmB += dc_speed_inc
        else:
            dc_pwmB = dc_max
    else:
        dc_pwmA = dc_speed_inc
        dc_pwmB = dc_speed_inc

    motorA_01_state = GPIO.LOW
    motorA_02_state = GPIO.HIGH
    motorB_01_state = GPIO.LOW
    motorB_02_state = GPIO.HIGH

    updateMotors()

    last_dc_pwmA = dc_pwmA
    last_dc_pwmB = dc_pwmB
    lastCall = FORWARD

# makes the robot go backward 
def backward():
    """Make the robot keep moving backward"""
    print("backward")

    if lastCall == BACKWARD: 
        #Speed incrementing
        if dc_pwmA + dc_speed_inc <= dc_max: 
            dc_pwmA += dc_speed_inc
        else:
            dc_pwmA = dc_max
        
        if dc_pwmB + dc_speed_inc <= dc_max: 
            dc_pwmB += dc_speed_inc
        else:
            dc_pwmB = dc_max
    else:
        dc_pwmA = dc_speed_inc
        dc_pwmB = dc_speed_inc

    motorA_01_state = GPIO.HIGH
    motorA_02_state = GPIO.LOW
    motorB_01_state = GPIO.HIGH
    motorB_02_state = GPIO.LOW

    updateMotors()

    last_dc_pwmA = dc_pwmA
    last_dc_pwmB = dc_pwmB
    lastCall = BACKWARD

##################### TURNING METHODS ####################

# turning LEFT means moving motor A faster than motor B 
def left():
    """Make the robot keep turning left"""
    print("left")

    if dc_pwmA + dc_inc <= dc_max: 
        dc_pwmA += dc_inc
    else
        dc_pwmA = dc_max
        dc_pwmB = dc_max - dc_inc

    updateMotors()

# turning RIGHT means moving motor B faster than motor A 
def right():
    """Make the robot keep turning right"""
    print("right")

    if dc_pwmB + dc_inc <= dc_max: 
        dc_pwmB += dc_inc
    else
        dc_pwmB = dc_max
        dc_pwmA = dc_max - dc_inc

    updateMotors()

##################### STOPPING METHODS ####################

# Stops all motors and sets PWM to 0
def stop():
    """Stop the robot"""
    print("stop")

    dc_pwmA = dc_min
    dc_pwmB = dc_min

    motorA_01_state = GPIO.LOW
    motorA_02_state = GPIO.LOW
    motorB_01_state = GPIO.LOW
    motorB_02_state = GPIO.LOW

    updateMotors() 

    last_dc_pwmA = dc_pwmA
    last_dc_pwmB = dc_pwmB
    lastCall = STOP

# Puts robot back into original FORWARD, BACKWARD, or STOP state
def stop_turning():
    """Stop turning"""
    print("stop turning")

    if lastCall == FORWARD: 
        motorA_01_state = GPIO.LOW
        motorA_02_state = GPIO.HIGH
        motorB_01_state = GPIO.LOW
        motorB_02_state = GPIO.HIGH
    elif lastCall == BACKWARD:
        motorA_01_state = GPIO.HIGH
        motorA_02_state = GPIO.LOW
        motorB_01_state = GPIO.HIGH
        motorB_02_state = GPIO.LOW
        
    dc_pwmA = last_dc_pwmA
    dc_pwmB = last_dc_pwmB

    updateMotors()

# Cleans up pins and pinmodes. Call init() again to start again. 
def terminate(): 
    stop()
    GPIO.output(stdby, GPIO.LOW)
    pwmA.stop() # stop PWM
    pwmB.stop() # stop PWM
    GPIO.cleanup() # cleanup all GPIO

# Helper method to update the pins to the set global variables
def updateMotors(): 
    pwmA.ChangeDutyCycle(dc_pwmA)
    pwmB.ChangeDutyCycle(dc_pwmB)

    GPIO.output(motorA_01, motorA_01_state)
    GPIO.output(motorA_02, motorA_02_state)
    GPIO.output(motorB_01, motorB_01_state)
    GPIO.output(motorB_02, motorB_02_state)

    
    

File: client/controls.c
#include <stdio.h>
#include "controls.h"

#include <wiringPi.h> 
#include <softPwm.h> 
#include <errno.h>

//////////////// VARIABLE DEFINITIONS ////////////////

// Motor Driver Pin declarations (Broadcom pinouts)
#define motorA_pwm 12 //GPIO12 (Pin 32) for Motor A PWM speed control
#define motorA_01 5 //GPIO05 (Pin 29) for Motor A 01 input
#define motorA_02 6 //GPIO06 (Pin 31) for Motor A 02 input
#define motorB_01 13 //GPIO13 (Pin 33) for Motor B 01 input
#define motorB_02 19 //GPIO19 (Pin 35) for Motor B 02 input
#define motorB_pwm 26 //GPIO26 (Pin 37) for Motor B PWM speed control
#define stdby 21 //GPIO21 (Pin 40) for Standby pin on Motor Driver
#define servo 18 //GPIO18 (Pin 12) for camera tilt

// Last call states 
#define FORWARD 1
#define BACKWARD 2
#define RIGHT 3
#define LEFT 4
#define STOP 0

// PWM Constants
#define PWM_MIN 0 
#define PWM_MAX 100 
#define PWM_SPINC 60 //how high the motors go when FORWARD or BACKWARD
#define PWM_INC 20 //for turning
#define PWM_SERVO_MAX 300
#define PWM_SINC 5

// Default pin modes
int pwmA_val = PWM_MIN;
int pwmB_val = PWM_MIN;
int last_pwmA_val = PWM_MIN;
int last_pwmB_val = PWM_MIN;
int motorA_01_state = LOW;
int motorA_02_state = LOW;
int motorB_01_state = LOW;
int motorB_02_state = LOW;
int pwmServo_val = PWM_MIN;

// Default lastCall to STOP
int lastCall = STOP;

void updateMotors(void);

/////////////////// INITIALIZER ///////////////////
// initializes wiringPi and pins
int init()
{
    printf("%s\n", "Initializing...");

    if(getuid()!=0) //wiringPi requires root privileges  
    {  
        printf("Error: wiringPi must be run as root.\n");  
        return 1;  
    } 

    wiringPiSetupGpio(); // Initializes wiringPi using the Broadcom GPIO pin numbers
    
    // not sure if we need soft pwm, but we'll try it first
    if (softPwmCreate(motorA_pwm, PWM_MIN, PWM_MAX) != 0) {
        printf("Motor A PWM failed to create %s\n", strerror(errno));
        return 1; 
    } 
    if (softPwmCreate(motorB_pwm, PWM_MIN, PWM_MAX) != 0) {
        printf("Motor B PWM failed to create %s\n", strerror(errno));
        return 1; 
    } 

    pinMode(motorA_01, OUTPUT); 
    pinMode(motorA_02, OUTPUT);
    pinMode(motorB_01, OUTPUT);
    pinMode(motorB_02, OUTPUT);
    pinMode(servo, PWM_OUTPUT);
    
    pwmSetMode(PWM_MODE_MS); 
    pwmSetClock(384); //clock at 50kHz (20us tick)
    pwmSetRange(1000); //range at 1000 ticks (20ms)

    pinMode(stdby, OUTPUT); 
    digitalWrite(stdby, HIGH); //Pull Standby high 

    stop();
    
    return 0; 
}

//////////////////// UP and DOWN Tilting methods //////////////

// increments the camera tilt servo by SINC val
void up() {
    if(pwmServo_val + PWM_SINC <= PWM_MAX) {
        pwmServo_val += PWM_SINC
    } else {
        pwmServo_val = PWM_MAX; 
    }
    printf("Up was called. Servo val: %d\n", pwmServo_val);
    pwmWrite(servo, pwmServo_val);
}

// decrements the camera tilt servo by SINC val
void down() {
    if(pwmServo_val - PWM_SINC >= PWM_MIN) {
        pwmServo_val -= PWM_SINC
    } else {
        pwmServo_val = PWM_MIN; 
    }
    printf("Down was called. Servo val: %d\n", pwmServo_val);
    pwmWrite(servo, pwmServo_val);
}

//////////////////// FORWARD and BACKWARD METHODS /////////////

// makes the robot go forward 
void forward()
{
    printf("%s\n", "forward");

    pwmA_val = PWM_SPINC; 
    pwmB_val = PWM_SPINC; 

    motorA_01_state = LOW;
    motorA_02_state = HIGH;
    motorB_01_state = LOW;
    motorB_02_state = HIGH;

    updateMotors();

    last_pwmA_val = pwmA_val;
    last_pwmB_val = pwmB_val;
    lastCall = FORWARD;
}

// makes the robot go backward 
void backward()
{
    printf("%s\n", "backward");

    pwmA_val = PWM_SPINC; 
    pwmB_val = PWM_SPINC; 

    motorA_01_state = HIGH;
    motorA_02_state = LOW;
    motorB_01_state = HIGH;
    motorB_02_state = LOW;

    updateMotors();

    last_pwmA_val= pwmA_val;
    last_pwmB_val = pwmB_val;
    lastCall = BACKWARD;
    
}

// turning LEFT means moving motor A faster than motor B 
void left()
{
    printf("%s\n", "left");

    if (lastCall == FORWARD) { 
        forward();
        pwmA_val = PWM_SPINC + PWM_INC;
        pwmB_val = PWM_SPINC - PWM_INC;
    }
    else if (lastCall == BACKWARD) {
        backward();
        pwmA_val = PWM_SPINC + PWM_INC;
        pwmB_val = PWM_SPINC - PWM_INC;
    }
    else {
        pwmA_val = PWM_INC * 2;
        pwmB_val = PWM_MIN;

        //go forward
        motorA_01_state = LOW;
        motorA_02_state = HIGH;
        motorB_01_state = LOW;
        motorB_02_state = HIGH;
    }

    updateMotors();
}

// turning RIGHT means moving motor B faster than motor A 
void right()
{
    printf("%s\n", "right");

    if (lastCall == FORWARD) { 
        forward();
        pwmA_val = PWM_SPINC - PWM_INC;
        pwmB_val = PWM_SPINC + PWM_INC;
    }
    else if (lastCall == BACKWARD) {
        backward();
        pwmA_val = PWM_SPINC - PWM_INC;
        pwmB_val = PWM_SPINC + PWM_INC;
    }
    else {
        pwmA_val = PWM_MIN;
        pwmB_val = PWM_INC * 2;

        //go forward
        motorA_01_state = LOW;
        motorA_02_state = HIGH;
        motorB_01_state = LOW;
        motorB_02_state = HIGH;
    }

    updateMotors();
}

///////////////////// STOPPING METHODS //////////////////

// Stops all motors and sets PWM to 0
void stop()
{
    printf("%s\n", "stop");

    pwmA_val = PWM_MIN;
    pwmB_val = PWM_MIN;

    motorA_01_state = LOW;
    motorA_02_state = LOW;
    motorB_01_state = LOW;
    motorB_02_state = LOW;

    updateMotors(); 

    last_pwmA_val= pwmA_val;
    last_pwmB_val = pwmB_val;
    lastCall = STOP;
}

// Puts robot back into original FORWARD, BACKWARD, or STOP state
void stop_turning()
{
    printf("%s\n", "stop_turning");

    if (lastCall == FORWARD) { 
        forward();
    }
    else if (lastCall == BACKWARD) {
        backward();
    }
    else {
        stop();
    }
}

// Helper method to update the pins to the set global variables
void updateMotors() 
{
    softPwmWrite(motorA_pwm, pwmA_val); 
    softPwmWrite(motorB_pwm, pwmB_val); 

    digitalWrite(motorA_01, motorA_01_state);
    digitalWrite(motorA_02, motorA_02_state);
    digitalWrite(motorB_01, motorB_01_state);
    digitalWrite(motorB_02, motorB_02_state);

}
